csrc/fusion_segmenter.cpp

```diff
@@ -4346,63 +4346,223 @@ void SegmentCandidateFinder::findSegments() {
+      // segments/squeeze ops that may appear in the final segmenter output.
+      // More details of the issue regarding merging horizontal groups can be
+      // found in issue 3829 -- https://github.com/NVIDIA/Fuser/issues/3829.
+      // Even with a squeeze op with 2 uses, this test case:
+      // https://github.com/NVIDIA/Fuser/blob/70ab277c7d91bcc24cd50dd75cedd79863a24f96/tests/python/test_python_frontend.py#L3666C1-L3666C30
+      // demonstrates that privatizing the squeeze op leads to horizontal groups
+      // that can't be merged back.
+      if (maybe_upcast_squeeze_out_tv->definition()->isA<SqueezeOp>() &&
+          maybe_upcast_squeeze_out_tv->uses().size() > 2) {
         continue;
       }
 
-      // Check if there's multiple uses of the upcast output
-      auto uses_of_upcast_out_tv = maybe_upcast_out_tv->uses();
-      if (uses_of_upcast_out_tv.size() < 2) {
+      // Check if there's multiple uses of the upcast/squeeze output
+      auto uses_of_out_tv = maybe_upcast_squeeze_out_tv->uses();
+      if (uses_of_out_tv.size() < 2) {
         continue;
       }
 
-      // If this is the first use of the upcast output, keep it as is
-      if (expr == uses_of_upcast_out_tv.front()) {
+      // If this is the first use of the upcast/squueze output, keep it as is
+      if (expr == uses_of_out_tv.front()) {
         continue;
       }
 
-      TensorView* upcast_out_tv_clone = castOp(
-          maybe_upcast_out_tv->dtype(),
-          maybe_upcast_op->input(0)->as<TensorView>());
+      TensorView* out_tv_clone = nullptr;
+      if (maybe_upcast_squeeze_out_tv->definition()->isA<UnaryOp>()) {
+        auto upcast_op =
+            maybe_upcast_squeeze_out_tv->definition()->as<UnaryOp>();
+        out_tv_clone = castOp(
+            maybe_upcast_squeeze_out_tv->dtype(),
+            upcast_op->input(0)->as<TensorView>());
+
+      } else {
+        auto squeeze_op =
+            maybe_upcast_squeeze_out_tv->definition()->as<SqueezeOp>();
+        out_tv_clone = squeeze(
+            squeeze_op->input(0)->as<TensorView>(),
+            squeeze_op->getSqueezeDimFlags(),
+            needs_to_squeeze_expanded(
+                squeeze_op->input(0)->as<TensorView>(),
+                squeeze_op->getSqueezeDimFlags()));
+      }
+
       TransformReplay::selfReplay(
-          maybe_upcast_out_tv->domain(), upcast_out_tv_clone->domain());
-      expr = ir_utils::replaceValInExprInputs(
-          expr, maybe_upcast_out_tv, upcast_out_tv_clone);
+          maybe_upcast_squeeze_out_tv->domain(), out_tv_clone->domain());
 
-      privatized_upcast_ops_[maybe_upcast_op].insert(
-          upcast_out_tv_clone->definition()->as<UnaryOp>());
+      auto new_expr = ir_utils::replaceValInExprInputs(
+          expr, maybe_upcast_squeeze_out_tv, out_tv_clone);
+
+      // make sure we don't have stale ops in privatized_ops_.
+      update_privatized_ops(privatized_ops_, expr, new_expr);
```

Comment:
Can you give some examples where this matters?

Reply:
Let's say

In_0 -> Cast -> In_1 -> Squeeze -> Use0, Use1

Firse, we'll privatize Squeeze, this will give us:

In_0 -> cast -> In_1 -> Squeeze -> Use0
                     -> Squeeze1 -> Use1


At this point, the privatize ops, holds the pair: Squeeze, Squeeze1

Next when we go to privatize cast, we will create:

In_0 -> cast -> In_1 -> Squeeze -> Use0
     -> cast -> In_2 -> Squeeze1' -> Use0

While doing this we will create a new Squeeze1` invalidating the previous Squeeze1.
Thus this is required.